Add program migration functionality

DennisInSky opened the pull request:
Resolves #2055

DennisInSky commented on gclient/src/api/calls.rs:
```diff
@@ -204,6 +204,74 @@ impl GearApi {
+        // example). Do we want to override it? If yes, then we have to check
+        // migration is done to another node (network). How?
+        // Another solution might be that we crate a brand new ProgramId which would
+        // allow us to migrate program even withing the same node/network
```
What are your thoughts about this? Either I have to make a check that we migrate to another node or I can generate a brand-new ProgramId and check if the code if already there then skip its migration and just clone the program

DennisInSky replied:
As far as I am aware, this call is meant to be used for testing purposes by our community when they want to take a program from test/prod net and bring it into their dev net (#2055).

breathx replied:
In case of existance program on target destination just return error
It will require querying by program id at the begging of the function

breathx replied:
> I assume this `migrate_program` call is designed for users, the problem is, users are not allowed to run the sudo calls

This fn will be used for local node, where sudo calls exist

> program migration maybe require us to add a new function in pallet-gear that allows us to submit program with state

Unbelievable api, we won't add something like that cause otherwise any smart-contract platforms become meaningless

DennisInSky commented on gsdk/src/storage.rs:
```diff
@@ -64,7 +64,7 @@ impl Api {
 
     /// Get program pages from program id.
-    pub async fn program_pages(&self, pid: H256) -> Result<types::GearPages> {
+    pub async fn program_pages(&self, pid: ProgramId) -> Result<types::GearPages> {
```
I think it is better to have more specific types whenever it is possible rather than abstract H256 or arrays

DennisInSky replied:
What kind of efficiency do we get here when we use `H256` in lieu of `ProgramId`. I do not see any except for confusion.

breathx replied:
I see no efficiency improvements to use H256 as well
The only thing that for user lib we need use gear_core::ProgramId, while for gsdk we can go with metadata_generated::ProgramId

DennisInSky replied:
I would go with `gear_core::*` in gsdk as well as this dependency is already there and as @clearloop metioned it will allow us to get rid of at least some types duplicated in metadata which in turn will minimize maintenance burden

DennisInSky commented on gsdk/src/storage.rs:
```diff
@@ -172,7 +172,7 @@ impl Api {
 impl Api {
     /// Get `InstrumentedCode` by `code_hash`
-    pub async fn code_storage(&self, code_hash: [u8; 32]) -> Result<InstrumentedCode> {
+    pub async fn gcode(&self, code_hash: [u8; 32]) -> Result<InstrumentedCode> {
```
Renamed for better consistency - we have had gprog and gpages thus gcode

DennisInSky replied:
Alright, I got the idea and will rename it back.

DennisInSky commented on gsdk/src/signer/calls.rs:
```diff
@@ -19,16 +19,27 @@
 };
 use anyhow::anyhow;
 use async_recursion::async_recursion;
 use gear_core::ids::{CodeId, MessageId, ProgramId};
```
Inconsistent usage of IDs throughout this crate confuses me:
- in some places it uses crate::metadata::gear_core::ids::{ProgramId, etc} whereas in this place it uses gear_core::{ProgramId, etc}
- at the same time the crate exposes some methods which use this similar abstractions from different crates (see comment below)
Why can't we use `gear_core::ids` everywhere? Why do we need `crate::metadata::gear_core::ids` at all? It causes a lot of confusing conversions.

DennisInSky replied:
Right, I think we have to crate another issue and adjust the code so it makes use of `gear_core` as much as possible. Taking account of the fact this code is for our eco-system, the code has to be nice, clean and consistent as everything publicly facing

breathx replied:
Previously we had auto generated extrinsics and calls which were using metadata::ids::*
Due to the fact that we have now dynamic api, seems reasonable to use gear_core::ids everywhere (furthermore, publicly reimport it in gsdk/gclient maybe?)

DennisInSky commented on gsdk/src/signer/calls.rs:
```diff
@@ -194,10 +205,102 @@ impl Signer {
+    pub async fn set_gprog(
+        &self,
+        program_id: ProgramId,
+        program: ActiveProgram,
```
This is an example where `ProgramId` comes from the `gear_core` crate whereas `ActiveProgram` comes from this crate. At the same time the complimentary `gprog` method uses both `ProgramId` and `ActiveProgram` from this crate

SabaunT reviewed (commented):
The solution reminds me of our past pallet tests, when we mocked a lot state transitions, by setting directly some values to storage instead of calling extrinsics, which will do this for us the proper way. If you emulate in one function logic done by a set of processed extrinsics, it increases complexity of the code maintenance.
Also the functionality is questionable for complicated test-cases, for example when the storage of the program is not completely changed by some message in waitlist.

DennisInSky commented:
> The solution reminds me of our past pallet tests, when we mocked a lot state transitions, by setting directly some values to storage instead of calling extrinsics, which will do this for us the proper way. If you emulate in one function logic done by a set of processed extrinsics, it increases complexity of the code maintenance. Also the functionality is questionable for complicated test-cases, for example when the storage of the program is not completely changed by some message in waitlist.

I am not going to defend this solution, but as far as I get, this feature is not meant to be used for prod and its main purpose is to allow contract developers to migrate a program from test/prod into their own dev environment for some testing. If we'd had extrinsics allowing uploading a program with its state, I would've definitely used them. The question about delayed messages (messages sitting in waitlist) is still to be a subject for discussion as in general there is more general issue - different timeline (chain height) on source and target networks. @breathx mentioned that we can "move clocks" on the target node into future to make up for the source one. In addition to this we can prohibit migration if the target node is in the future compared to the source one.

breathx commented on gsdk/src/storage.rs:
```diff
@@ -181,12 +181,22 @@ impl Api {
     }
 
+    /// Get `InstrumentedCode` length by its `code_hash`
+    pub async fn code_len_storage(&self, code_hash: [u8; 32]) -> Result<u32> {
```
```suggestion
    pub async fn code_len_storage(&self, code_id: CodeId) -> Result<u32> {
```

breathx commented on gsdk/src/signer/calls.rs:
```diff
@@ -194,6 +205,101 @@ impl Signer {
+                    subxt::dynamic::Value::u128(*program_page.0 as u128),
+                ],
+            );
+            let page_buf_inner = gear_core::memory::PageBufInner::try_from(program_page.1.clone())
+                .map_err(|_| Error::PageBroken(*program_page.0, program_id.encode_hex()))?;
+            let value = gear_core::memory::PageBuf::from_inner(page_buf_inner);
```
```suggestion
            let page_buf_inner = PageBufInner::try_from(program_page.1.clone())
                .map_err(|_| Error::PageBroken(*program_page.0, program_id.encode_hex()))?;
            let value = PageBuf::from_inner(page_buf_inner);
```
Import types directly and recheck if here proper constructor for such cases

DennisInSky replied:
@breathx I have imported the types, but I am not sure what you mean by `proper ctor` - there are no ctors for these types which take an array, there are only conversion methods which I use here

breathx commented on gcli/src/cmd/program.rs:
```diff
@@ -74,10 +74,10 @@ impl Program {
 
         // Get program
-        let program = api.gprog(pid).await?;
+        let program = api.gprog(ProgramId(pid.0)).await?;
```
Lets deal with core ids

breathx reviewed (changes requested):
Requesting changes to prevent merging before solving all of the conversations properly.

Otherwise, lgtm 👍🏻

breathx commented:
> The solution reminds me of our past pallet tests, when we mocked a lot state transitions, by setting directly some values to storage instead of calling extrinsics, which will do this for us the proper way. If you emulate in one function logic done by a set of processed extrinsics, it increases complexity of the code maintenance.
> Also the functionality is questionable for complicated test-cases, for example when the storage of the program is not completely changed by some message in waitlist.

You're right, seems like our good old tests :D
But we had a lot of conversation with ecosystem and decided that we don't need any stuff related to waitlist and bns (at least for now)

breathx commented:
@DennisInSky Just realised that there is one important thing to be also added within the PR: migrate GasTree nodes representing reservations inside the program (specified in ActiveProgram structure). Don't forget to reserve proper amount of balance from GasNode origins to avoid post-execution panics. Test for reservation sending within the PR would be also good!

DennisInSky commented on gclient/tests/node.rs:
```diff
@@ -1,44 +1,128 @@
     // salt to the same node twice
-    upload_program_to_node(&node_1, WASM_BINARY, &salt).await;
-    upload_program_to_node(&node_2, WASM_BINARY, &salt).await;
+    upload_program_to_node(&node_1, demo_mul_by_const::WASM_BINARY, &salt, Some(42u64)).await;
```
This won't work for the used dapp. It requires initialization payload

DennisInSky commented:
> @DennisInSky Just realised that there is one important thing to be also added within the PR: migrate GasTree nodes representing reservations inside the program (specified in ActiveProgram structure). Don't forget to reserve proper amount of balance from GasNode origins to avoid post-execution panics. Test for reservation sending within the PR would be also good!

@breathx I think I have addressed all the comments in this PR except for this one. Do you really want me to do this as a part of this PR or I can do it in the next one. Anyway, I have some questions about it so I would rather merge this code in.

shamilsan commented on gclient/src/api/calls.rs:
```diff
@@ -204,6 +204,78 @@ impl GearApi {
+    pub async fn migrate_program(
+        &self,
+        src_program_id: ProgramId,
+        tgt_node_api: &GearApi,
```
Usually, we use pair `src`/`dest`.
```suggestion
        dest_node_api: &GearApi,
```

shamilsan commented on gclient/src/api/calls.rs:
```diff
@@ -204,6 +204,78 @@ impl GearApi {
+        let src_code_len = self
+            .0
+            .api()
+            .code_len_storage(src_program.code_hash.0.into())
```
It is better to get code hash one time and keep it in a variable.

NikVolf commented:
Programs should preserve their id-s I think (or have a way to preserve)

Otherwise interacting programs will break

DennisInSky commented:
> Programs should preserve their id-s I think (or have a way to preserve)
> 
> Otherwise interacting programs will break

It's been preserved. There is even assertion of this fact in the test.

breathx reviewed (approved):
Waiting further discussed improvements 👍🏻

shamilsan commented on gclient/src/api/calls.rs:
```diff
@@ -204,6 +204,72 @@ impl GearApi {
+        src_program_id: ProgramId,
+        dest_node_api: &GearApi,
+    ) -> Result<ProgramId> {
+        // TODO: Ideally we have to check program is not already there (via gprog, for
+        // example). Do we want to override it? If yes, then we have to check
+        // migration is done to another node (network). How?
+        // Another solution might be that we crate a brand new ProgramId which would
+        // allow us to migrate program even withing the same node/network
```
Have discussed, need to add an issue and reference to it here if the check will not be added to this PR.

DennisInSky replied:
@shamilsan , @breathx I have fixed it in this PR. Please check this commit https://github.com/gear-tech/gear/pull/2331/commits/02f9036677434f3f733dbeef5ee1ff3e1fb6d49d. Found and fixed one bug while was doing this.

shamilsan commented on gclient/src/api/calls.rs:
```diff
@@ -204,6 +206,90 @@ impl GearApi {
+        let dest_program_id = src_program_id;
+
+        // Collect data from the source program
+        let src_free_balance = self.free_balance(src_program_id).await.map_or_else(
+            |e| {
+                if let Error::GearSDK(GsdkError::StorageNotFound) = e {
+                    Ok(0)
+                } else {
+                    Err(e)
+                }
+            },
+            |v| Ok(v),
+        )?;
```
Could we use `or_else` here?
```suggestion
        let src_free_balance = self.free_balance(src_program_id).await.or_else(
            |e| {
                if let Error::GearSDK(GsdkError::StorageNotFound) = e {
                    Ok(0)
                } else {
                    Err(e)
                }
            })?;
```
